
<robot name="underwater" xmlns:xacro="http://www.ros.org/wiki/xacro">

    <!-- * * * Color Definitions * * * -->

    <material name="white">
        <color rgba="1 1 1 1"/>
    </material>

    <material name="green2">
        <color rgba="0.8 0.9 0.8 0.9"/>
    </material> 

    <material name="red">
        <color rgba="0.8 0 0 1"/>
    </material>

    <material name="gray">
        <color rgba="0.5 0.5 0.5 0.9"/>
    </material>

    <material name="transparent">
        <color rgba="1 1 1 0.4"/>
    </material>

    <material name="full_transparent">
        <color rgba="1 1 1 0"/>
    </material>

    <!-- * * * Link Definitions * * * -->

    <!-- mesh files -->
	<link name="robot_base_link">
        
        <inertial>
            <origin xyz="0 0 0" rpy="0 0 0"/>
            <mass value="5.99" />
            <inertia ixx="0.08319857" ixy="0.0008657224" ixz="0.0009579166" iyy="0.08357262" iyz="0.0001832366" izz="0.12186"/>
        </inertial>
        
        <collision>
            <origin xyz="0 0 .01" rpy="0 0 0"/>
            <geometry>
                <cylinder radius="0.25" length="0.11"/>
            </geometry>
        </collision>

        <visual>
            <origin rpy="0 0 -2.094" xyz="0 0 -.0976"/>
            <geometry>
                <mesh filename="package://underwater_sim/models/meshes/robot/MainBody.dae" scale="0.001 0.001 0.001"/>
            </geometry>
            <material name="white"/>
        </visual>
    </link>

    <link name="center_link">
        <inertial>
            <origin xyz="0 0 0" rpy="0 0 0"/>
            <mass value="0.0" />
            <inertia ixx="0" ixy="0.0" ixz="0.0" iyy="0" iyz="0.0" izz="0"/>
        </inertial>
        <visual>
            <origin rpy="0 0 0" xyz="-0.0302 -0.03215 -0.018"/>
            <geometry>
                <mesh filename="package://underwater_sim/models/meshes/robot/CentralPropeller.dae" scale="0.001 0.001 0.001"/>
            </geometry>
            <material name="black"/>
        </visual>
    </link> 


    <link name="shaft1_link">
        <visual>
            <origin rpy="0 0 3.1415" xyz="0 -.09 -.034"/>
            <geometry>
                <mesh filename="package://underwater_sim/models/meshes/robot/Shaft.dae" scale="0.001 0.001 0.001"/>
            </geometry>
            <material name="white"/>
        </visual>
    </link> 

    <link name="shaft2_link">
        <visual>
            <origin rpy="0 0 3.1415" xyz="0 -.09 -.034"/>
            <geometry>
                <mesh filename="package://underwater_sim/models/meshes/robot/Shaft.dae" scale="0.001 0.001 0.001"/>
            </geometry>
            <material name="white"/>
        </visual>
    </link>  

    <link name="shaft3_link">
        <visual>
            <origin rpy="0 0 3.1415" xyz="0 -.09 -.034"/>
            <geometry>
                <mesh filename="package://underwater_sim/models/meshes/robot/Shaft.dae" scale="0.001 0.001 0.001"/>
            </geometry>
            <material name="white"/>
        </visual>
    </link> 

    <link name="cylinder1_link">
        <collision>
            <origin xyz="0 0 .12" rpy="0 0 0"/>
            <geometry>
                <cylinder radius="0.045" length="0.09"/>
            </geometry>
        </collision>
        <visual>
            <origin rpy="0 0 0" xyz="0 0 -.014"/>
            <geometry>
                <mesh filename="package://underwater_sim/models/meshes/robot/Cylinder.dae" scale="0.001 0.001 0.001"/>
            </geometry>
            <material name="transparent"/>
        </visual>
        
    </link> 
    <link name="cylinder2_link">
        <collision>
            <origin xyz="0 0 .12" rpy="0 0 0"/>
            <geometry>
                <cylinder radius="0.045" length="0.09"/>
            </geometry>
        </collision>
        <visual>
            <origin rpy="0 0 0" xyz="0 0 -.014"/>
            <geometry>
                <mesh filename="package://underwater_sim/models/meshes/robot/Cylinder.dae" scale="0.001 0.001 0.001"/>
            </geometry>
            <material name="transparent"/>
        </visual>
    </link>

    <link name="cylinder3_link">
        <collision>
            <origin xyz="0 0 .12" rpy="0 0 0"/>
            <geometry>
                <cylinder radius="0.045" length="0.09"/>
            </geometry>
        </collision>
        <visual>
            <origin rpy="0 0 0" xyz="0 0  -.014"/>
            <geometry>
                <mesh filename="package://underwater_sim/models/meshes/robot/Cylinder.dae" scale="0.001 0.001 0.001"/>
            </geometry>
            <material name="transparent"/>
        </visual>
    </link> 

    <link name="camera_link">
        <collision>
            <origin xyz="-0.028 0 0" rpy="0 1.5708 0"/>
            <geometry>
                <mesh filename="package://underwater_sim/models/meshes/robot/camera.dae" scale="0.0015 0.0015 0.0015"/>
            </geometry>
        </collision>
        <visual>
            <origin xyz="-0.028 0 0" rpy="0 1.5708 0" />
            <geometry>
                <mesh filename="package://underwater_sim/models/meshes/robot/camera.dae" scale="0.0015 0.0015 0.0015"/>
            </geometry>
        </visual>
    </link> 

    <link name="imu_link">
        <collision>
            <origin xyz="0 0 0.057" rpy="0 0 0"/>
            <geometry>
                <mesh filename="package://underwater_sim/models/meshes/robot/imu.dae" scale="0.0012 0.0012 0.0012"/>
            </geometry>
        </collision>
        <visual>
            <origin xyz="0 0 0.057" rpy="0 0 0" />
            <geometry>
                <mesh filename="package://underwater_sim/models/meshes/robot/imu.dae" scale="0.0012 0.0012 0.0012"/>
            </geometry>
        </visual>
    </link> 
    
    <link name="motorcylinder1_link">

        <visual>
            <origin rpy="0 0 0" xyz="0 .244 0.00"/>
            <geometry>
                <mesh filename="package://underwater_sim/models/meshes/robot/TubeMotor.dae" scale="0.001 0.001 0.001"/>
            </geometry>
            <material name="transparent"/>
        </visual>
    </link> 

    <link name="motorcylinder2_link">

        <visual>
            <origin rpy="0 0 -2.094" xyz=".2114 -.122 0"/>
            <geometry>
                <mesh filename="package://underwater_sim/models/meshes/robot/TubeMotor.dae" scale="0.001 0.001 0.001"/>
            </geometry>
            <material name="transparent"/>
        </visual>
    </link> 

    <link name="motorcylinder3_link">

        <visual>
            <origin rpy="0 0 2.094" xyz="-.2114 -.122 0"/>
            <geometry>
                <mesh filename="package://underwater_sim/models/meshes/robot/TubeMotor.dae" scale="0.001 0.001 0.001"/>
            </geometry>
            <material name="transparent"/>
        </visual>
    </link>

    <link name="wheel1_link">
        
        <inertial>
            <origin xyz="0 0 0" rpy="0 0 0"/>
            <mass value="0.605" />
            <inertia ixx="0.0007561942" ixy="0.00" ixz="0.00" iyy="0.001441249" iyz="0.00" izz="0.0007561973"/>
        </inertial>
        <collision>
            <origin rpy="1.5708 0 0" xyz="0 0 0"/>
            <geometry>
                <cylinder radius="0.08" length="0.03"/>
            </geometry>
        </collision>
    </link>

    <link name="wheel2_link">
        <inertial>
            <origin xyz="0 0 0" rpy="0 0 0"/>
            <mass value="0.605"/>
            <inertia ixx="0.0007561942" ixy="0.00" ixz="0.00" iyy="0.001441249" iyz="0.00" izz="0.0007561973"/>
        </inertial>
        <collision>
            <origin rpy="1.5708 0 0" xyz="0 0 0"/>
            <geometry>
                <cylinder radius="0.08" length="0.03"/>
            </geometry>
        </collision>
    </link>


    <link name="wheel3_link">
        <inertial>
            <origin xyz="0 0 0" rpy="0 0 0"/>
            <mass value="0.605"/>
            <inertia ixx="0.0007561942" ixy="0.0" ixz="0.0" iyy="0.001441249" iyz="0.00" izz="0.0007561973"/>
        </inertial>
        <collision>
            <origin rpy="1.5708 0 0" xyz="0 0 0"/>
            <geometry>
                <cylinder radius="0.08" length="0.03"/>
            </geometry>
        </collision>
    </link>  

    <link name="cerlab_logo_link">

        <visual>
            <origin rpy="0 0 3.1415" xyz="0 0.15 0.065"/>
            <geometry>
                <mesh filename="package://underwater_sim/models/meshes/robot/cerlabLogo.dae" scale="0.00075 0.00075 0.00075"/>
            </geometry>
            <material name="gray"/>
        </visual>
    </link> 


	
<!-- * * * Joint Definitions * * * -->

    <joint name="center_motor" type="continuous">
        <parent link="robot_base_link"/>
        <child link="center_link"/>
        <origin xyz="0.000 0 0.0000" rpy="0 0 0"/>
        <limit lower="-1" upper="1" effort="0.0" velocity="0.00"/>
        <axis xyz="0 0 1"/>
    </joint>

    <joint name="motor1" type="continuous">
        <parent link="robot_base_link"/>
        <child link="shaft1_link"/>
        <origin xyz="0 0.245 0" rpy="0 0 0"/>
        <limit lower="-1" upper="1" effort="0.0" velocity="0.00"/>
        <axis xyz="0 1.5708 0"/>
    </joint>  

    <joint name="motor2" type="continuous">
        <parent link="robot_base_link"/>
        <child link="shaft2_link"/>
        <origin xyz="-.21223 -0.1224 0" rpy="0 0 2.094"/>
        <limit lower="-1" upper="1" effort="0.0" velocity="0.00"/>
        <axis xyz="0 1 0 "/>
    </joint> 
  
    <joint name="motor3" type="continuous">
        <parent link="robot_base_link"/>
        <child link="shaft3_link"/>
        <origin xyz=".21223 -0.1224 0" rpy="0 0 -2.094"/>
        <limit lower="-1" upper="1" effort="0.0" velocity="0.00"/>
        <axis xyz="0 1 0"/>
    </joint>  

    <joint name="wheel1_joint" type="fixed">
        <parent link="shaft1_link"/>
        <child link="wheel1_link"/>
        <origin xyz="0 .117 0" rpy="0 0 0"/>
        <limit lower="-1" upper="1" effort="0.0" velocity="0.00"/>
        <axis xyz="0 1 0"/>
   </joint>

         
    <joint name="wheel2_joint" type="fixed">
        <parent link="shaft2_link"/>
        <child link="wheel2_link"/>
        <origin xyz="0 .117 0" rpy="0 0 0"/>
        <limit lower="-1" upper="1" effort="0.0" velocity="0.00"/>
        <axis xyz="0 1 0"/>
   </joint>

    <joint name="wheel3_joint" type="fixed">
        <parent link="shaft3_link"/>
        <child link="wheel3_link"/>
        <origin xyz="0 .117 0" rpy="0 0 0"/>
        <limit lower="-1" upper="1" effort="0.0" velocity="0.00"/>
        <axis xyz="0 1 0"/>
    </joint>

    <joint name="cylinder1_joint" type="fixed">
        <parent link="robot_base_link"/>
        <child link="cylinder1_link"/>
        <origin rpy="0 0 0" xyz=".1048 .0605 -.00"/>
        <limit lower="-1" upper="1" effort="0.0" velocity="0.00"/>
        <axis xyz="0 1 0"/>
    </joint>

    <joint name="cylinder2_joint" type="fixed">
        <parent link="robot_base_link"/>
        <child link="cylinder2_link"/>
        <origin rpy="0 0 0" xyz="-.1048 .0605 -.00"/>
        <limit lower="-1" upper="1" effort="0.0" velocity="0.00"/>
        <axis xyz="0 1 0"/>
    </joint>

    <joint name="cylinder3_joint" type="fixed">
        <parent link="robot_base_link"/>
        <child link="cylinder3_link"/>
        <origin rpy="0 0 0" xyz="0 -.121 0"/>
        <limit lower="-1" upper="1" effort="0.0" velocity="0.00"/>
        <axis xyz="0 1 0"/>
    </joint>

    <joint name="camera_joint" type="fixed">
        <parent link="cylinder2_link"/>
        <child link="camera_link"/>
        <origin rpy="0 0 -1.5708" xyz="0 -0.03 0.1"/>
        <axis xyz="0 1 0"/>
    </joint>
    
    <joint name="imu_joint" type="fixed">
        <parent link="cylinder2_link"/>
        <child link="imu_link"/>
        <origin rpy="0 0 0" xyz="0 0 0"/>
        <axis xyz="0 1 0"/>
    </joint>
    
    <joint name="motorcylinder1_joint" type="fixed">
        <parent link="robot_base_link"/>
        <child link="motorcylinder1_link"/>
        <origin xyz="0 0 0" rpy="0 0 0"/>
        <limit lower="-1" upper="1" effort="0.0" velocity="0.00"/>
        <axis xyz="0 1 0"/>
    </joint>

    <joint name="motorcylinder2_joint" type="fixed">
        <parent link="robot_base_link"/>
        <child link="motorcylinder2_link"/>
        <origin xyz="0 0 0" rpy="0 0 0"/>
        <limit lower="-1" upper="1" effort="0.0" velocity="0.00"/>
        <axis xyz="0 1 0"/>
    </joint>

    <joint name="motorcylinder3_joint" type="fixed">
        <parent link="robot_base_link"/>
        <child link="motorcylinder3_link"/>
        <origin xyz="0 0 0" rpy="0 0 0"/>
        <limit lower="-1" upper="1" effort="0.0" velocity="0.00"/>
        <axis xyz="0 1 0"/>
    </joint>

    <joint name="cerlab_logo_joint" type="fixed">
        <parent link="robot_base_link"/>
        <child link="cerlab_logo_link"/>
        <origin xyz="0 0 0" rpy="0 0 0"/>
        <limit lower="-1" upper="1" effort="0.0" velocity="0.00"/>
        <axis xyz="0 1 0"/>
    </joint>


    <gazebo reference="cylinder1_link">
        <visual name='visual'>
        <material>
          <lighting>1</lighting>
          <script>
            <uri>file://media/materials/scripts/gazebo.material</uri>
            <name>Gazebo/Grey</name>
          </script>
          <ambient>0.3 0.3 0.3 0.4</ambient>
          <diffuse>0.7 0.7 0.7 0.4</diffuse>
          <specular>0.01 0.01 0.01 0.4</specular>
          <emissive>0 0 0 0.4</emissive>
          <shader type='vertex'>
            <normal_map>__default__</normal_map>
          </shader>
        </material>
        <cast_shadows>1</cast_shadows>
        <transparency>0.7</transparency>
        </visual>
    </gazebo>
    <gazebo reference="cylinder2_link">
            <visual name='visual'>
        <material>
          <lighting>1</lighting>
          <script>
            <uri>file://media/materials/scripts/gazebo.material</uri>
            <name>Gazebo/Grey</name>
          </script>
          <ambient>0.3 0.3 0.3 0.4</ambient>
          <diffuse>0.7 0.7 0.7 0.4</diffuse>
          <specular>0.01 0.01 0.01 0.4</specular>
          <emissive>0 0 0 0.4</emissive>
          <shader type='vertex'>
            <normal_map>__default__</normal_map>
          </shader>
        </material>
        <cast_shadows>1</cast_shadows>
        <transparency>0.7</transparency>
        </visual>
    </gazebo>
    <gazebo reference="cylinder3_link">
            <visual name='visual'>
        <material>
          <lighting>1</lighting>
          <script>
            <uri>file://media/materials/scripts/gazebo.material</uri>
            <name>Gazebo/Grey</name>
          </script>
          <ambient>0.3 0.3 0.3 0.4</ambient>
          <diffuse>0.7 0.7 0.7 0.4</diffuse>
          <specular>0.01 0.01 0.01 0.4</specular>
          <emissive>0 0 0 0.4</emissive>
          <shader type='vertex'>
            <normal_map>__default__</normal_map>
          </shader>
        </material>
        <cast_shadows>1</cast_shadows>
        <transparency>0.7</transparency>
        </visual>
    </gazebo>
    <gazebo reference="motorcylinder1_link">
        <visual name='visual'>
        <material>
          <lighting>1</lighting>
          <script>
            <uri>file://media/materials/scripts/gazebo.material</uri>
            <name>Gazebo/Grey</name>
          </script>
          <ambient>0.3 0.3 0.3 0.4</ambient>
          <diffuse>0.7 0.7 0.7 0.4</diffuse>
          <specular>0.01 0.01 0.01 0.4</specular>
          <emissive>0 0 0 0.4</emissive>
          <shader type='vertex'>
            <normal_map>__default__</normal_map>
          </shader>
        </material>
        <cast_shadows>1</cast_shadows>
        <transparency>0.7</transparency>
        </visual>
    </gazebo>
    <gazebo reference="motorcylinder2_link">
            <visual name='visual'>
        <material>
          <lighting>1</lighting>
          <script>
            <uri>file://media/materials/scripts/gazebo.material</uri>
            <name>Gazebo/Grey</name>
          </script>
          <ambient>0.3 0.3 0.3 0.4</ambient>
          <diffuse>0.7 0.7 0.7 0.4</diffuse>
          <specular>0.01 0.01 0.01 0.4</specular>
          <emissive>0 0 0 0.4</emissive>
          <shader type='vertex'>
            <normal_map>__default__</normal_map>
          </shader>
        </material>
        <cast_shadows>1</cast_shadows>
        <transparency>0.7</transparency>
        </visual>
    </gazebo>
    <gazebo reference="motorcylinder3_link">
            <visual name='visual'>
        <material>
          <lighting>1</lighting>
          <script>
            <uri>file://media/materials/scripts/gazebo.material</uri>
            <name>Gazebo/Grey</name>
          </script>
          <ambient>0.3 0.3 0.3 0.4</ambient>
          <diffuse>0.7 0.7 0.7 0.4</diffuse>
          <specular>0.01 0.01 0.01 0.4</specular>
          <emissive>0 0 0 0.4</emissive>
          <shader type='vertex'>
            <normal_map>__default__</normal_map>
          </shader>
        </material>
        <cast_shadows>1</cast_shadows>
        <transparency>0.7</transparency>
        </visual>
    </gazebo>
    
    <gazebo>
        <plugin name="BuoyancyPlugin" filename="libBuoyancyPlugin.so">
            <fluid_density>1000</fluid_density>
            <link name="robot_base_link">
                <center_of_volume>0 0 0</center_of_volume>
                <volume>0.00599</volume>
            </link>
        </plugin>
    </gazebo>
<!--
    <gazebo>
        <plugin name="BuoyancyPlugin" filename="libBuoyancyPlugin.so">
            <fluid_density>1000</fluid_density>fluid_density>
            <link name="wheel1_link">
                <center_of_volume>0 0 0</center_of_volume>
            </link>
        </plugin>
    </gazebo>

    <gazebo>
        <plugin name="BuoyancyPlugin" filename="libBuoyancyPlugin.so">
            <fluid_density>1000</fluid_density>fluid_density>
            <link name="wheel2_link">
                <center_of_volume>0 0 0</center_of_volume>
            </link>
        </plugin>
    </gazebo>
    <gazebo>
        <plugin name="BuoyancyPlugin" filename="libBuoyancyPlugin.so">
            <fluid_density>1000</fluid_density>fluid_density>
            <link name="wheel3_link">
                <center_of_volume>0 0 0</center_of_volume>
            </link>ls
        </plugin>
    </gazebo>
-->

    <gazebo>
        <plugin name="LiftDragPlugin" filename="libLiftDragPlugin.so">
            <air_density>1000</air_density><cla>0</cla><cla_stall>0</cla_stall>
            <cda>0</cda><cda_stall>0</cda_stall><alpha_stall>0</alpha_stall><a0>0</a0>
            <area>0.5654866776</area>
            <upward>0 0 1</upward>
            <forward>1 0 0</forward>
            <link_name>robot_base_link</link_name>
            <cp>0 0 0</cp>
        </plugin>
    </gazebo>
    
    <!-- camera -->
     <gazebo reference="camera_link">
       	<sensor type="camera" name="camera1">
         	<update_rate>30.0</update_rate>
         	<camera name="head">
                <pose>0 0 0 0 0 0</pose> 
                <horizontal_fov>1.3962634</horizontal_fov>
                <image>
                    <width>800</width>
                    <height>800</height>
                    <format>R8G8B8</format>
                </image>
                <clip>
                    <near>0.02</near>
                    <far>300</far>
                </clip>
                <noise>
                    <type>gaussian</type>
                    <mean>0.0</mean>
                    <stddev>0.007</stddev>
                </noise>
         	</camera>
         	<plugin name="camera_controller" filename="libgazebo_ros_camera.so">
           		<alwaysOn>true</alwaysOn>
           		<updateRate>0.0</updateRate>
           		<cameraName>underwater/camera1</cameraName>
           		<imageTopicName>image_raw</imageTopicName>
           		<cameraInfoTopicName>camera_info</cameraInfoTopicName>
           		<frameName>camera_link</frameName>
           		<hackBaseline>0.07</hackBaseline>
           		<distortionK1>0.0</distortionK1>
           		<distortionK2>0.0</distortionK2>
           		<distortionK3>0.0</distortionK3>
           		<distortionT1>0.0</distortionT1>
           		<distortionT2>0.0</distortionT2>
         	</plugin>
       	</sensor>
     </gazebo>
    
    <!--
  	<gazebo reference="imu_link">
    	<gravity>true</gravity>
    	<sensor name="imu_sensor" type="imu">
      		<always_on>true</always_on>
      		<update_rate>100</update_rate>
      		<visualize>true</visualize>
      		<topic>gazebo_imu</topic>
      		<plugin filename="libgazebo_ros_imu_sensor.so" name="imu_plugin">
        		<topicName>imu</topicName>
        		<bodyName>imu_link</bodyName>
        		<updateRateHZ>10.0</updateRateHZ>
        		<gaussianNoise>0.0</gaussianNoise>
        		<xyzOffset>0 0 0</xyzOffset>
        		<rpyOffset>0 0 0</rpyOffset>
        		<frameName>imu_link</frameName>
      		</plugin>
      		<pose>0 0 0 0 0 0</pose>
    	</sensor>
  	</gazebo>
    -->
    <gazebo>
        <plugin name="fake_imu" filename="libfake_imu.so">
            <orientationType>quaternion</orientationType>
        </plugin>
    </gazebo>
    
    
    <gazebo>
        <plugin name="pose_sensor" filename="libposition.so">
            <robotNamespace>robot_base_link</robotNamespace>
        </plugin>
    </gazebo>
    
    <gazebo>
        <plugin name="encoder_sensor" filename="libencoder.so">
            <jointName>motor1</jointName>
            <topicName>encoder_1</topicName>
        </plugin>
    </gazebo>
        
    <gazebo>
        <plugin name="encoder_sensor" filename="libencoder.so">
            <jointName>motor2</jointName>
            <topicName>encoder_2</topicName>
        </plugin>
    </gazebo>
    
    <gazebo>
        <plugin name="encoder_sensor" filename="libencoder.so">
            <jointName>motor3</jointName>
            <topicName>encoder_3</topicName>
        </plugin>
    </gazebo>

</robot> 
